ampd/src/grpc/blockchain_service.rs

```diff
@@ -1,70 +1,90 @@
     ContractsResponse, QueryRequest, QueryResponse, SubscribeRequest, SubscribeResponse,
 };
 use async_trait::async_trait;
-use cosmrs::tendermint::block;
-use futures::{Stream, StreamExt};
-use tokio::sync::mpsc;
-use tokio::time;
-use tokio_stream::wrappers::ReceiverStream;
+use futures::{Stream, TryStreamExt};
+use report::LoggableError;
+use tokio_stream::StreamExt;
 use tonic::{Request, Response, Status};
+use tracing::error;
+use valuable::Valuable;
 
-pub struct Service {}
+use crate::event_sub;
+use crate::grpc::event_filters;
 
-impl Service {
-    pub fn new() -> Self {
-        Self {}
+pub struct Service<E>
+where
+    E: event_sub::EventSub,
+{
+    event_sub: E,
+}
+
+impl<E> Service<E>
+where
+    E: event_sub::EventSub,
+{
+    pub fn new(event_sub: E) -> Self {
+        Self { event_sub }
     }
 }
 
 #[async_trait]
-impl BlockchainService for Service {
-    type SubscribeStream =
-        Pin<Box<dyn Stream<Item = Result<SubscribeResponse, Status>> + Send + 'static>>;
+impl<E> BlockchainService for Service<E>
+where
+    E: event_sub::EventSub + Send + Sync + 'static,
+{
+    type SubscribeStream = Pin<Box<dyn Stream<Item = Result<SubscribeResponse, Status>> + Send>>;
 
     async fn subscribe(
         &self,
-        _req: Request<SubscribeRequest>,
+        req: Request<SubscribeRequest>,
     ) -> Result<Response<Self::SubscribeStream>, Status> {
-        // TODO: replace the dummy implementation
-        let (tx, rx) = mpsc::channel(10000);
-
-        tokio::spawn(async move {
-            let mut height: block::Height = 10u32.into();
-            let duration = Duration::from_secs(3);
-            let mut interval = time::interval(duration);
-
-            loop {
-                let _ = tx
-                    .send(ampd_proto::subscribe_response::Event::BlockBegin(
-                        ampd_proto::EventBlockBegin {
-                            height: height.value(),
-                        },
-                    ))
-                    .await;
-                interval.tick().await;
-
-                let _ = tx
-                    .send(ampd_proto::subscribe_response::Event::BlockEnd(
-                        ampd_proto::EventBlockEnd {
-                            height: height.value(),
-                        },
-                    ))
-                    .await;
-                interval.tick().await;
-
-                height = height.increment();
-            }
-        });
+        let SubscribeRequest {
+            filters,
+            include_block_begin_end,
+        } = req.into_inner();
+
+        let filters: event_filters::EventFilters = (filters, include_block_begin_end)
+            .try_into()
+            .map_err(|err| {
+                error!(
+                    err = LoggableError::from(&err).as_value(),
+                    "invalid event filters provided for event subscription"
+                );
```

Comment:
I don't like logging and returning the error, because that makes it confusing who should be responsible for it. Wherever the error is finally handled should the logging happen